hacks/prune-rhcos-snapshots/prune-rhcos-snapshots.py

```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/env python3
+import os
+
+import boto3
+import openshift as oc
+import json
+import yaml
+import urllib3
+import pathlib
+import subprocess
+
+
+def create_recycle_bin_rule(client, res_type: str):
+    response = client.create_rule(
+        RetentionPeriod={
+            'RetentionPeriodValue': 60,
+            'RetentionPeriodUnit': 'DAYS'
+        },
+        Description='string',
+        ResourceType=res_type,
+    )
+    print(f'Created new retention rule: {response["Identifier"]}')
+
+
+if __name__ == '__main__':
+
+    gopath = pathlib.Path(os.getenv('GOPATH'))
+    if not gopath.is_dir():
+        print('Set GOPATH env before running')
+        exit(1)
+
+    installer_git_path = gopath.joinpath('src/github.com/openshift/installer')
+    if not installer_git_path.is_dir():
+        print(f'Requires recent clone of github.com/openshift/installer in GOPATH: {str(installer_git_path)}')
+        exit(1)
+
+    http = urllib3.PoolManager()
+
+    must_preserve = set()  # entries with "region:ami" that we know we should preserve. Used for sanity checking results.
+    will_preserve = set()  # entries in the analysis marked with preserve
+
+    # File used as cache for release information and the AMIs mentioned by their installer commit
+    riu_path = pathlib.Path('rhcos_in_use.yaml')
+    rhcos_in_use = dict()
+    if riu_path.exists():
+        rhcos_in_use = yaml.safe_load(riu_path.read_text())
+
+    arches = ['amd64', 's390x', 'ppc64le', 'arm64']
+    for arch in arches:
+        rc_url = f'https://{arch}.ocp.releases.ci.openshift.org/graph'
+        r = http.request('GET', rc_url)
+        if r.status != 200:
+            print(f'Error retrieving release controller graph data: {r}')
+            exit(1)
+
+        graph_str = r.data.decode('utf-8')
+        graph = json.loads(graph_str)
+
+        for node in graph['nodes']:
+            payload_pullspec = node['payload']
+            payload_version = node['version']
+            payload_key = f'{arch}:{payload_version}'
+
+            if payload_key in rhcos_in_use:
+                # We've already cached this payload RHCOS version
+                print(f'Skipping {payload_key} since it is cached')
+                continue
+
+            print(f"Processing: {payload_version}  / {payload_pullspec}")
+            for retry in (1, 2, 3, 4):
+                release_json_result = oc.invoke('adm', cmd_args=['release', 'info', payload_pullspec, '-o=json'],
+                                                no_namespace=True, auto_raise=False)
+                if release_json_result.status() == 0:
+                    break
+                else:
+                    print(f'Error acquiring image information: {release_json_result.err()}')
+
+            if release_json_result.status() != 0:
+                if '.ci' in payload_version or '.nightly' in payload_version:
+                    continue
```

Comment:
Actually... we don't want to `preserve` everything in a nightly for the indefinite future, but we _do_ want to preserve it for long enough to become a promoted release. So should there also just be a blanket exemption for AMIs until they're old enough (say, a month)? Presumably garbage collection config would cover that, perhaps that's why it's not built in here.

Reply:
Right - If we ran this again the future, older nightlies would have been garbage collected from the release controller. As written, this tool would not try to preserve them.